server/manager/src/main/java/org/apache/accumulo/manager/upgrade/UpgradeProgressTracker.java

```diff
@@ -0,0 +1,162 @@
+          AccumuloDataVersion.get());
+      Preconditions.checkArgument(newVersion > rootVersion,
+          "New Root version (%s) must be greater than current Root version (%s)", newVersion,
+          rootVersion);
+      Preconditions.checkArgument(newVersion <= zooKeeperVersion,
+          "New Root version (%s) expected to be <= ZooKeeper version (%s)", newVersion,
+          zooKeeperVersion);
+      Preconditions.checkArgument(newVersion > metadataVersion,
+          "New Root version (%s) must be greater than current Metadata version (%s)", newVersion,
+          metadataVersion);
+      rootVersion = newVersion;
+      put(context, this);
+    }
+
+    public int getMetadataVersion() {
+      return metadataVersion;
+    }
+
+    public synchronized void updateMetadataVersion(ServerContext context, int newVersion)
+        throws KeeperException, InterruptedException {
+      Objects.requireNonNull(context, "ServerContext must be supplied");
+      Preconditions.checkArgument(newVersion <= AccumuloDataVersion.get(),
+          "New version (%s) cannot be larger than current data version (%s)", newVersion,
+          AccumuloDataVersion.get());
+      Preconditions.checkArgument(newVersion > metadataVersion,
+          "New Metadata version (%s) must be greater than current version (%s)", newVersion,
+          metadataVersion);
+      Preconditions.checkArgument(newVersion <= zooKeeperVersion,
+          "New Metadata version (%s) expected to be <= ZooKeeper version (%s)", newVersion,
+          zooKeeperVersion);
+      Preconditions.checkArgument(newVersion <= rootVersion,
+          "New Metadata version (%s) expected to be <= Root version (%s)", newVersion, rootVersion);
+      metadataVersion = newVersion;
+      put(context, this);
+    }
+  }
+
+  private static String getZPath(ServerContext context) {
+    return context.getZooKeeperRoot() + Constants.ZUPGRADE_STATUS;
+  }
+
+  private static synchronized void put(ServerContext context, ComponentVersions cv)
+      throws KeeperException, InterruptedException {
+    Objects.requireNonNull(context, "ServerContext must be supplied");
+    Objects.requireNonNull(cv, "ComponentVersions object  must be supplied");
+    final String zpath = getZPath(context);
+    final ZooReaderWriter zrw = context.getZooSession().asReaderWriter();
+    zrw.sync(zpath);
+    if (!zrw.exists(zpath)) {
+      zrw.mkdirs(zpath);
+    }
+    zrw.putPersistentData(zpath, GSON.get().toJson(cv).getBytes(UTF_8), NodeExistsPolicy.OVERWRITE);
+  }
+
+  public static synchronized ComponentVersions get(ServerContext context)
+      throws KeeperException, InterruptedException {
+    final String zpath = getZPath(context);
+    final int currentVersion = AccumuloDataVersion.getCurrentVersion(context);
+    final ZooReaderWriter zrw = context.getZooSession().asReaderWriter();
+    zrw.sync(zpath);
+    if (!zrw.exists(zpath)) {
```

Comment:
I think this code will change a bit if my other suggestions are followed. But, you don't need to sync here. Instead, you can attempt to create the node (since most of the time, that's the usual case). Creating the node requires quorum, so you don't have to worry about calling sync, and it protects you better, since the client could become unsynced immediately after asking it to sync. If creating the node fails because it already exists, then you can parse it and determine whether you should resume or abort, based on what's there.

Reply:
Addressed in 7d19a91